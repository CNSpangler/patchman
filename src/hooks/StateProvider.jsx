import React, { createContext, useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';

const StateProvider = ({ children }) => {
  const [url, setUrl] = useState('');
  const [body, setBody] = useState('');
  const [method, setMethod] = useState('GET');
  const [results, setResults] = useState('');
  const [history, setHistory] = useState([]);

  const handleUrlChange = ({ target }) => {
    setUrl(target.value);
  };
  
  const handleMethodChange = ({ target }) => {
    setMethod(target.value);
  };
  
  const handleBodyChange = ({ target }) => {
    setBody(target.value);
  };
  
  const handleSubmit = (e) => {
    e.preventDefault();
  
    makeRequest(url, method, body)
      .then(response => setResults(response));
  
    // dependent state change
    setHistory(prevHistory => ([
      ...prevHistory,
      {
        url,
        method,
        body
      }
    ]));
  };

  return (
    <StateContext.Provider value={{ url, body, method, results, history, handleUrlChange, handleMethodChange, handleBodyChange, handleSubmit }}>
      {children}
    </StateContext.Provider>
  );
};

StateProvider.propTypes = {
  children: PropTypes.node
};

export default StateProvider;
