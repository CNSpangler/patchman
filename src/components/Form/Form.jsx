import React from 'react';
import styles from '../App/App.css';
import { useUrl, useHandleUrlChange, useHandleBodyChange, useBody, useHandleMethodChange, useHandleSubmit, useMethod } from '../../hooks/StateProvider';

/* eslint-disable react/prop-types */
const RadioGroup = ({ name, onChange, children }) => {
  const radioButtonsWithNameAndOnChange = React.Children.map(children, child => {
    return React.cloneElement(child, {
      name,
      onChange
    });
  });

  return (
    <>
      {radioButtonsWithNameAndOnChange}
    </>
  );
};

const RadioButton = ({ name, value, onChange }) => (
  <div className={styles.radioButtonContainer} >
    <input id={value} type="radio" name={name} value={value} onChange={onChange} />
    <label htmlFor={value} className={styles.radioButton}>{value}</label>
  </div>
);

/* eslint-enable react/prop-types */
const Form = () => {
  const url = useUrl();
  const onUrlChange = useHandleUrlChange();
  const body = useBody();
  const onBodyChange = useHandleBodyChange();
  const onMethodChange = useHandleMethodChange();
  const onSubmit = useHandleSubmit();

  return (
    <form onSubmit={onSubmit} className={styles.Form}>
      <RadioGroup name="method" onChange={onMethodChange} className={styles.RadioGroup}>
        <RadioButton value="POST" />
        <RadioButton value="GET" />
        <RadioButton value="PATCH" />
        <RadioButton value="PUT" />
        <RadioButton value="DELETE" />
      </RadioGroup>
      <input type="text" value={url} placeholder="URL" onChange={onUrlChange} className={styles.url} />
      <input type="textarea" value={body} placeholder="Raw JSON Body" onChange={onBodyChange} className={styles.body} />
      <div className={styles.headers}>Placeholder for headers</div>
      <button className={styles.button}>Submit</button>
    </form>
  );
};

export default Form;
