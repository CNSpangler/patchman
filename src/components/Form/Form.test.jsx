import React from 'react';
import { shallow } from 'enzyme';
import Form from './Form.jsx';
import { StateProvider } from '../../hooks/StateProvider.jsx';

describe('Form component', () => {
  it('renders Form ', () => {
    const wrapper = shallow(
      <StateProvider>  
        <Form />
      </StateProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
