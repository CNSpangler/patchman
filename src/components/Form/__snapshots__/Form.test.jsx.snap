// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component renders Form  1`] = `
<ContextProvider
  value={
    Object {
      "body": "",
      "handleBodyChange": [Function],
      "handleMethodChange": [Function],
      "handleSubmit": [Function],
      "handleUrlChange": [Function],
      "history": Array [],
      "method": "GET",
      "results": "",
      "url": "",
    }
  }
>
  <Form />
</ContextProvider>
`;
